## plot2DO: Plot 2D Occupancies

**plot2DO** is an R package for computing and visualizing the two-dimensional (2D) occupancies. Instead of using the typical one-dimensional (1D) occupancy/coverage, obtained by stacking all the mapped reads regardless of their lengths, it is more informative to compute the relative occupancy given by DNA fragments of specific lengths in a matrix form.

## Download
To download this package you can use the Github user interface and just click the **Download** button. If you want to download the package from the terminal, then you need to install first a *`git`* client of your choice, using the package manager that is available for your system. For example, in Ubuntu or other Debian-based distributions of Linux, you can use *`apt-get`*:
```
$ sudo apt-get install git
```
, while in Fedora, CentOS, or Red Hat Linux, you can use *`yum`*:
```
$ sudo yum install git
```
Installers for OSX (Mac) and Windows (PC) are available for download at the Git websites: *`http://git-scm.com/download/mac`* and *`http://git-scm.com/download/win`*.
After *`git`* has been installed, run the following command from the folder where you want to download the **plot2DO** package:
```
$ git clone https://github.com/rchereji/plot2DO.git
```


## Dependencies
**plot2DO** uses the following R packages: *`caTools, colorRamps, GenomicAlignments, GenomicRanges, optparse, rtracklayer`*. To install these packages, open R and execute the following commands:
```{r}
> # Install caTools, colorRamps and optparse packages from CRAN:
> install.packages(c("caTools", "colorRamps", "optparse"))
>                  
> # Get the latest version of Bioconductor:
> source("https://bioconductor.org/biocLite.R")
> biocLite()
> 
> # Install the remaining Bioconductor packages:
> biocLite(c("GenomicAlignments", "GenomicRanges", "rtracklayer"))
```


## Usage
After the R packages have been installed, you can execute **plot2DO** from *`bash`*. To get to the help page, open a terminal and run:
```
$ Rscript plot2DO.R --help
Usage: plot2DO.R [options]


Options:
        -f FILE, --file=FILE
                Dataset file name [options: BED or BAM format]

        -r REFERENCE, --reference=REFERENCE
                Reference points to align [options: TSS (default), TTS, Plus1]

        -l MINLENGTH, --minLength=MINLENGTH
                The smallest DNA fragment to be considered [default = 50]

        -L MAXLENGTH, --maxLength=MAXLENGTH
                The largest DNA fragment to be considered [default = 200]

        -u UPSTREAM, --upstream=UPSTREAM
                Length of the upstream region that will be plotted [default = 1000]

        -d DOWNSTREAM, --downstream=DOWNSTREAM
                Length of the downstream region that will be plotted [default = 1000]

        -h, --help
                Show this help message and exit
```
The only required input is the name of the file that contains the aligned genomic data (*e.g.* *`data.bam, data.bed`*). To generate the default 2D occupancy plot (around TSS, all DNA fragments with the length L between 50 - 200 bp, 2kb windows centered on TSS), run the following command:
```
$ Rscript plot2DO.R --file data.bam
```
This is equivalent to either of the extended forms:
```
$ Rscript plot2DO.R --file data.bam --reference TSS --minLength 50 --maxLength 200 --upstream 1000 --downstream 1000
$ Rscript plot2DO.R -f data.bed -r TSS -l 50 -L 200 -u 1000 -d 1000
```


## Outputs
**plot2DO** will create a new folder called **2D_occupancy**, which will contain the 2D occupancy, saved in *`RData`* and *`csv`* formats, and the plots, saved in a *`pdf`* format. An example figure generated by **plot2DO** (Dataset: *`439J_3_H4_25U_IP`*; Reference: TSS; 50 bp ≤ L ≤ 200 bp; Upstream/Downstream window = 1 kb) is shown below:
![](misc/img.png)


## Extra functions
The proper minimum (*`--minLength`*) and maximum (*`--maxLength`*) fragment sizes to be displayed in the 2D occupancy plot can be obtained by inspecting the histogram of all DNA fragment lengths that are present in the sample. This can be done using the additional function **computeLengthHistogram**. The available options are shown below:
```
$ Rscript computeLengthHistogram.R --help
Usage: computeLengthHistogram.R [options]


Options:
        -f FILE, --file=FILE
                Dataset file name [options: BED or BAM format]

        -o OUTPUTS, --outputs=OUTPUTS
                Types of outputs to be generated, separated by commas only [options: pdf, csv, RData; default = pdf,csv,RData]

        -h, --help
                Show this help message and exit

```


## Citation
If you use this program in your research, please cite our article. (not available yet)


## License
**plot2DO** is freely available under the MIT License.
